service/history/engine/engineimpl/record_child_execution_completed.go

```diff
@@ -75,22 +75,24 @@ func (e *historyEngineImpl) RecordChildExecutionCompleted(
 					)
 					return workflow.ErrStaleState
 				}
+				// This can happen if the child execution is already completed, so it's deleted from mutable state
+				// We must return a non-retryable error to avoid infinite retries
 				return &types.EntityNotExistsError{Message: "Pending child execution not found."}
 			}
 			if ci.StartedID == constants.EmptyEventID {
-				if startedID >= mutableState.GetNextEventID() {
-					e.metricsClient.IncCounter(metrics.HistoryRecordChildExecutionCompletedScope, metrics.StaleMutableStateCounter)
-					e.logger.Error("Encounter stale mutable state in RecordChildExecutionCompleted",
-						tag.WorkflowDomainName(domainEntry.GetInfo().Name),
-						tag.WorkflowID(workflowExecution.GetWorkflowID()),
-						tag.WorkflowRunID(workflowExecution.GetRunID()),
-						tag.WorkflowInitiatedID(initiatedID),
-						tag.WorkflowStartedID(startedID),
-						tag.WorkflowNextEventID(mutableState.GetNextEventID()),
-					)
-					return workflow.ErrStaleState
-				}
-				return &types.EntityNotExistsError{Message: "Pending child execution not started."}
+				// This means that the child execution is initiated but not started in the view of the parent workflow,
```

Comment:
do we have sufficient child workflow integration tests to confidently say this is safe?

Reply:
This is hard to trigger in an integration tests environment. This only happens when the following 3 conditions are met:
1. There is a shard movement of the parent workflow, and the shard is moved to a different history host.
2. The child workflow still send request to the old host.
3. The child workflow was initiated from the parent workflow when the parent workflow was in the old host, but the transfer task was processed after shard movement by the new host.

Reply:
Makes sense